### PR TITLE
feat(chowly-client): add initial chowly integration

### PR DESCRIPTION
**Pull request by KanoniPan**

### CODE REVIEW

File: bin/prepublish.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+distPackage.types = './index.d.ts';
+delete distPackage.scripts;
+delete distPackage.devDependencies;
+delete distPackage.config;
+delete distPackage.husky;
+delete distPackage.files;
+delete distPackage.directories;
+delete distPackage['lint-staged'];
+
+writeFileSync('./dist/package.json', JSON.stringify(distPackage, null, '  '));
```
Comment (strdr4605) [inline bin/prepublish.js]:
```suggestion
writeFileSync('./dist/package.json', JSON.stringify(distPackage, null, 2));
```

File: package.json [hunk 2]
```diff
@@ -0,0 +1,86 @@
+{
+    "name": "chowly",
```
Comment (strdr4605) [inline package.json]: Maybe we should pick another name and also use the namespace.
```suggestion
    "name": "@goparrot/chowly-api-client",
```

File: src/client/chowly-client.api.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import { plainToClass } from 'class-transformer';
+import { IConfiguration } from '../interface';
+import { Sections } from '../model/chowly.model';
+import { sleep } from '../utils';
+
+export class ChowlyClient {
+    private readonly client: AxiosInstance;
+    private readonly apiKey: string;
+
+    private basePrefix: string = '/api/toos/v2';
```
Comment (strdr4605) [inline src/client/chowly-client.api.ts]: What does `toos` mean?

Comment (KanoniPan) [reply src/client/chowly-client.api.ts]: its chowly related prefix

File: package.json (hunk 2 above)
Comment (KanoniPan) [reply package.json]: was inspired by google, they have almost the same naming for this kind of features

File: src/client/chowly-client.api.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import axios, { AxiosInstance, AxiosRequestConfig } from 'axios';
+import { plainToClass } from 'class-transformer';
+import { IConfiguration } from '../interface';
+import { Sections } from '../model/chowly.model';
+import { sleep } from '../utils';
+
+export class ChowlyClient {
```
Comment (strdr4605) [inline src/client/chowly-client.api.ts]: Maybe it makes sense to rename class as `ApiClient` as it already in **chowly** package/namespace?

File: package.json (hunk 2 above)
Comment (strdr4605) [reply package.json]: Then to follow the convention of @goparrot packages, we can use `@goparrot/chowly`.

File: src/client/chowly-client.api.ts (hunk 4 above)
Comment (KanoniPan) [reply src/client/chowly-client.api.ts]: btw inside **square-connect-plus** we have SquareClient, i think we need one-two more opinions about that

File: .github/ISSUE_TEMPLATE/1-bug-report.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
+---
```
Comment (strdr4605) [inline .github/ISSUE_TEMPLATE/1-bug-report.md:1]: Too many changes for one PR. 

Issue/PR templates and workflows could be done in 2 separate PRs

File: src/client/chowly-client.api.ts (hunk 4 above)
Comment (strdr4605) [reply src/client/chowly-client.api.ts]: I was looking at the original `square-connect`

```
var SquareConnect = require('square-connect');
var defaultClient = SquareConnect.ApiClient.instance;
```

and looks like they use `ApiClient`

Comment (KanoniPan) [reply src/client/chowly-client.api.ts]: ok, i see, but they have `SquareConnect` so user know that it comes from Square, anyway lets wait for one more review and if we get the most votes on `ApiClient` i will change it ;)

File: .nycrc.json [hunk 6]
```diff
@@ -0,0 +1,16 @@
+{
+  "//": "read https://github.com/istanbuljs/nyc#configuring-nyc",
+  "check-coverage": true,
+  "per-file": true,
+  "statements": [70, 100],
+  "branches": [30, 100],
```
Comment (dragosh1011) [inline .nycrc.json]: is this the current coverage value?

File: package.json [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        "publish": "npm run build:dist && node bin/prepublish.js && npm publish dist",
+        "publish:dev": "npm run publish --tag dev",
+        "publish:dev:dry": "npm run publish:dev --dry-run",
+        "version": "echo \"use 'npm run release'\" && exit 1",
+        "release": "standard-version && git push && git push --tags && npm run publish && npm run github-release",
+        "release:dry": "npm run publish:dev:dry && standard-version --dry-run"
+    },
+    "author": "Panevschi Ruslan",
+    "license": "MIT",
+    "devDependencies": {
```
Comment (dragosh1011) [inline package.json:61]: add `axios` and `class-transformer` to `peerDependencies` specifying the minimum acceptable versions

File: .nycrc.json (hunk 6 above)
Comment (KanoniPan) [reply .nycrc.json]: current coverage is 100%

File: src/client/chowly-client.api.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
+import axios, { AxiosInstance, AxiosRequestConfig } from 'axios';
```
Comment (dragosh1011) [inline src/client/chowly-client.api.ts]: remove `.api` prefix or rename cholwy-api-client

File: src/client/chowly-client.api.ts (hunk 4 above)
Comment (dragosh1011) [reply src/client/chowly-client.api.ts]: I am for `ChowlyApiClient`

File: src/client/chowly-client.api.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            },
+        );
+
+        this.client.interceptors.response.use(
+            undefined,
+            async (error): Promise<any> => {
+                if (!error.config.metadata) {
+                    error.config.metadata = { retryCount: 1 };
+                }
+                console.log('err', error);
```
Comment (dragosh1011) [inline src/client/chowly-client.api.ts]: remove `console.log`

File: src/client/chowly-client.api.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+                if (error.response?.status >= 500 && error.config.metadata.retryCount < this.config.maxRetries!) {
+                    error.config.metadata.retryCount++;
+                    await sleep(this.config.retryDelay! + this.config.retryDelay! * 0.2 * Math.random());
```
Comment (dragosh1011) [inline src/client/chowly-client.api.ts]: why you need `Math.random()`?

**Review by dragosh1011 — CHANGES_REQUESTED** [review]

File: .nycrc.json (hunk 6 above)
Comment (dragosh1011) [reply .nycrc.json]: update values in this file to 100% then

File: src/client/chowly-client.api.ts (hunk 10 above)
Comment (KanoniPan) [reply src/client/chowly-client.api.ts]: some sort of delay, still thinking how to implement this correctly, or to use exponentialDelay from google, or to make something like this with given milliseconds + some kind of random

**Review by dragosh1011 — APPROVED** [review]

File: bin/prepublish.js [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+distPackage.main = './index.js';
+distPackage.types = './index.d.ts';
+delete distPackage.scripts;
+delete distPackage.devDependencies;
+delete distPackage.config;
+delete distPackage.husky;
+delete distPackage.files;
+delete distPackage.directories;
+delete distPackage['lint-staged'];
+
```
Comment (dragosh1011) [inline bin/prepublish.js:16]: add `delete distPackage.jest`

File: src/interface/chowly.interface.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
+export interface IConfiguration {
```
Comment (dragosh1011) [inline src/interface/chowly.interface.ts]: separate configuration interface and menu interfaces in different files

File: src/interface/chowly.interface.ts [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    subsection_id: string;
+    title: string;
+}
+
+export interface IItem {
+    nutritional_info: INutritionalInfo;
+    customizations: ICustomization[];
+    item_id: string;
+    title: string;
+    image_url: string | null | undefined;
```
Comment (dragosh1011) [inline src/interface/chowly.interface.ts]: `image_url?: string;`

File: package.json [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+        "remark-github": "^9.0.0",
+        "remark-lint-emphasis-marker": "^2.0.0",
+        "remark-lint-strong-marker": "^2.0.0",
+        "standard-version": "^8.0.0",
+        "ts-jest": "^25.5.1",
+        "ts-node": "^8.10.1",
+        "tslint-config-prettier": "^1.18.0",
+        "typescript": "^3.8.3"
+    },
+    "jest": {
```
Comment (strdr4605) [inline package.json]: Jest settings are getting big.
Maybe is better to extract in `jest.config.js`?

Comment (KanoniPan) [reply package.json]: sure i will extract it

File: README.md [hunk 15]
```diff
@@ -1 +1,70 @@
...
+    $ npm i @goparrot/chowly-api-client
+
+## Usage
+
+### Simple example
+
+```typescript
+import { ChowlyClient } from '@goparrot/chowly-api-client';
+
+(async () => {
```
Comment (ddimitrioglo) [inline README.md]: Do not use (async () => {})() if you are not using await inside it

File: package.json [hunk 16]
```diff
@@ -0,0 +1,97 @@
+{
+    "name": "@goparrot/chowly-api-client",
+    "version": "0.0.1",
+    "description": "Typescript library which implements Chowly API",
+    "engines": {
+        "node": ">=10"
```
Comment (ddimitrioglo) [inline package.json]: in `.nvmrc` you have set v12, I think you have to consolidate

File: package.json [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+        "pre-commit": "git add . && npm run format:staged && npm run lint && npm run coverage && npm run build:dist",
+        "prepublishOnly": "echo \"use 'npm run publish'\" && exit 1",
+        "publish": "npm run build:dist && node bin/prepublish.js && npm publish dist",
+        "publish:dev": "npm run publish --tag dev",
+        "publish:dev:dry": "npm run publish:dev --dry-run",
+        "version": "echo \"use 'npm run release'\" && exit 1",
+        "release": "standard-version && git push && git push --tags && npm run publish && npm run github-release",
+        "release:dry": "npm run publish:dev:dry && standard-version --dry-run"
+    },
+    "author": "Panevschi Ruslan",
```
Comment (ddimitrioglo) [inline package.json]: Everything you have done during working hours - is GoParrot's

File: package.json [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+        "deepmerge": "^3.2.0 || ^4.0.0",
+        "class-transformer": ">=0.2.0"
+    },
+    "devDependencies": {
+        "@commitlint/cli": "^8.3.5",
+        "@commitlint/config-conventional": "^8.3.4",
+        "@commitlint/travis-cli": "^8.3.5",
+        "@types/chai": "^4.2.11",
+        "@types/jest": "^25.2.1",
+        "@types/node": "^13.13.4",
```
Comment (ddimitrioglo) [inline package.json]: 13 is not LTS, better use ^12

File: src/client/chowly-api-client.ts [hunk 19]
```diff
@@ -0,0 +1,32 @@
+import { AxiosInstance, AxiosResponse } from 'axios';
+import { plainToClass } from 'class-transformer';
+import { IConfiguration, ISections } from '../interface';
+import { Sections } from '../model';
+import { createAxiosInstance } from '../utils';
+
+export class ChowlyApiClient {
+    readonly client: AxiosInstance;
+    private readonly config: IConfiguration;
+    private basePrefix: string = '/api/toos/v2';
```
Comment (ddimitrioglo) [inline src/client/chowly-api-client.ts:10]: If I'm not mistaken Axios has special option for this `prefix: '/api/toos/v2'` I guess

File: src/interface/configuration.interface.ts [hunk 20]
```diff
@@ -0,0 +1,5 @@
+export interface IConfiguration {
```
Comment (ddimitrioglo) [inline src/interface/configuration.interface.ts]: maybe `IClientConfig`?

File: src/utils/common.ts [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+
+export interface IFacadeConfig {
+    axiosConfig?: AxiosRequestConfig;
+    axiosRetryConfig?: IAxiosRetryConfig;
+    apiKey?: string;
+    maxRetries?: number;
+}
+
+/**
+ * @param {IFacadeConfig} config
```
Comment (ddimitrioglo) [inline src/utils/common.ts]: leave only `@throws`

Comment (KanoniPan) [reply src/utils/common.ts]: why leave only that? better then to remove this block

File: src/client/chowly-api-client.ts (hunk 19 above)
Comment (KanoniPan) [reply src/client/chowly-api-client.ts:10]: axios interface doesn't have prefix, i tried some varianties, but it seems like it doesn't work, the only thing that worked for me was url, but it overrides /menu path

**Review by ddimitrioglo — APPROVED** [review]

**Pull request merged**